src/app/reporting/Engine.cpp

```diff
@@ -658,6 +658,116 @@ bool Engine::MergeOverlappedAttributePath(const AttributePathParams & aAttribute
     });
 }
 
+bool Engine::ClearTombPaths()
+{
+    bool pathReleased = false;
+    mGlobalDirtySet.ForEachActiveObject([&](auto * path) {
+        if (path->mGeneration == 0)
+        {
+            mGlobalDirtySet.ReleaseObject(path);
+            pathReleased = true;
+        }
+        return Loop::Continue;
+    });
+    return pathReleased;
+}
+
+bool Engine::MergeDirtyPathsUnderSameCluster()
+{
+    mGlobalDirtySet.ForEachActiveObject([&](auto * outerPath) {
+        if (outerPath->HasWildcardEndpointId() || outerPath->HasWildcardClusterId() || outerPath->mGeneration == 0)
```

Comment:
Yes, the check for endpoint is redundant, however, based on spec 7.9 and 7.10, a cluster on an endpoint should be independent with the (same) cluster on another endpoint. Thus it is unlikely to have paths like E*/Cx in global dirty set although it might be wildly used in read requests.